LanguageFeatures/Extension-types/static_analysis_extension_types_A33_t03.dart

```diff
@@ -0,0 +1,47 @@
+// Copyright (c) 2024, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+/// @assertion An extension type V is always-exhaustive if and only if its
+/// instantiated representation type is always-exhaustive.
+///
+/// @description Check that an extension type is always-exhaustive if its
+/// representation type is always-exhaustive.
```

Comment:
In this situation, `B` isn't always-exhaustive, which means that `BET` also isn't always-exhaustive, which means that switch statements aren't subject to exhaustiveness analysis.

How would we detect that?

Presumably, we'd have a switch statement whose cases aren't exhaustive, and it should not be an error.

The function could have return type `void`, because we don't want it to be an error if the location after the switch statement is considered reachable (it _must_ be considered reachable, it would be a soundness issue if it is considered unreachable).

So the main point is that we can "measure" the status (is always-exhaustive vs. not) by having that non-exhaustive switch statement, and expecting no error if we expect not-always-exhaustiveness.

Reply:
OK. Checking that the issue reported in dart-lang/sdk#54444 has been fixed could be done by checking that no compile-time error is emitted for the following:

```dart
class B {}
extension type BET(B _) {}

String test1(BET b) { // "OK" would mean that the spec change has been implemented.
  switch (b) {
    case B _: return 'B';
  }
}
```

The error used to be emitted because `BET` wasn't erased to `B` during flow analysis, and then flow analysis wouldn't conclude anything useful from knowing that `case B` returns (and then we'd get the "implicitly returns null at end of function" error).

But my understanding was that this series of tests is about the opposite property: We already have other tests about using the erasure to reason about run-time type tests, is this series of tests we want to see that the erasure is not applied for any other reason.

Isn't it true that we already have tests testing that we _do_ get the erasure when reasoning about run-time type tests? (That's basically about switch statements being able or unable to complete normally.)

Reply:
Thanks!